sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -1746,37 +1769,108 @@ pi_result piProgramGetInfo(pi_program Program, pi_program_info ParamName,
   case PI_PROGRAM_INFO_DEVICES:
     return ReturnValue(Program->Context->Device);
   case PI_PROGRAM_INFO_BINARY_SIZES: {
-    size_t SzBinary = 0;
-    ZE_CALL(zeModuleGetNativeBinary(Program->ZeModule, &SzBinary, nullptr));
-    // This is an array of 1 element, initialize if it were scalar.
+    size_t SzBinary;
+    if (Program->State == _pi_program::IL) {
+      // The OpenCL spec indicates that PI_PROGRAM_INFO_BINARY_SIZES is not
+      // defined in this case, but it does not say what to do if it happens.
+      // Returning a zero size seems reasonable.
+      SzBinary = 0;
+    } else {
+      assert(Program->State == _pi_program::Object ||
+             Program->State == _pi_program::Exe ||
+             Program->State == _pi_program::LinkedExe);
```

Comment:
Every check for `LinkedExe` state goes with OR check for `Exe` state. I think it will simplify the changes if we just have single `Exe` state to represent both cases

Reply:
This is not true.  Look at the implementation of `ModuleIterator`.

Reply:
Admitted, but adding anew globally visible state to accommodate internal implementation of iterator is an overkill, in my view. We could check the number/properties of linked objects instead.

Reply:
The way I see it, `State` describes the shape of the data contained by the `_pi_program` object.  There are four possible shapes, so there are four possible states.  The `LinkedExe` state also helps a reader understand the code better.  Look at the places where `LinkedExe` shows up in the comments.  Those comments would all have to be changed if we removed this state.  Also look at the long comment documenting the `LinkedExe` state itself.  This comment would need to be moved to some less obvious place.

Practically speaking, we could of course replace `if (State == LinkedExe)` with something like `if ((State == Exe) && !LinkedPrograms.empty())`.  That seems longer and less clear to me, though.

Reply:
It's certainly a matter of taste, so I won't insist on changing this now. Please look at the failures reported by the CI. Otherwise changes look good to me.